R/driver-snowflake.R

```diff
@@ -5,6 +5,58 @@ NULL
 #' @rdname DBI-classes
 setClass("Snowflake", contains = "OdbcConnection")
 
+#' Connecting to Snowflake via ODBC
+#'
+#' @description
+#' ## `odbcConnectionColumns()`
+#'
+#' If the catalog, or the schema arguments are NULL, attempt to infer
+#' by querying for CURRENT_DATABASE() and CURRENT_SCHEMA().  We do this
+#' to aid with performance, as the SQLColumns method is more performant
+#' when restricted to a particular DB/schema.
+#' @inheritParams DBI::dbListFields
+#' @rdname Snowflake
+#' @usage NULL
+setMethod("odbcConnectionColumns_", c("Snowflake", "character"),
+  function(conn,
+           name,
+           ...,
+           catalog_name = NULL,
+           schema_name = NULL,
+           column_name = NULL,
+           exact = FALSE) {
+    if(is.null(catalog_name) || is.null(schema_name)) {
+      res <- dbGetQuery(conn, "SELECT CURRENT_DATABASE() AS CAT, CURRENT_SCHEMA() AS SCH")
+      if(is.null(catalog_name) && !is.na(res$CAT)) {
+        catalog_name <- res$CAT
+      }
+      if(is.null(schema_name) && !is.na(res$SCH)) {
+        schema_name <- res$SCH
+      }
+    }
+    callNextMethod(conn = conn, name = name, ..., catalog_name = catalog_name,
+      schema_name = schema_name, column_name = column_name, exact = exact)
+  }
+)
+
+#' @rdname Snowflake
+setMethod("dbExistsTableForWrite", c("Snowflake", "character"),
+  function(conn, name, ...) {
```

Comment:
I think schema and catalog should be the explicit arguments here since you're modifying them, and I'm pretty sure that should allow you to drop the `do.call`.

Reply:
Hey - that's also a good thought!

But unless you feel strongly about this, I would err on the side of leaving the signature of `dbExistsTableForWrite` the same as `dbExistsTable` - at least for now, with the current default.

Reply:
I think we can leave the signature of the generic the same, but it's ok for the signature of method to be different?

Reply:
Done - let me know if that's not what you had in mind.